### PR TITLE
Add OnBehalfOf authentication to V2 for edgehub

### PR DESCRIPTION
**Pull request by patilsnr**
Adds OnBehalfOf authentication to SDK for Edge devices

### CODE REVIEW

File: iothub/device/src/Authentication/ClientAuthenticationForEdgeHubOnBehalfOf.cs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    /// <summary>
+    /// Authentication method that generates shared access signature (SAS) token with refresh, based on a provided shared access key (SAK).
+    /// Build for gateway scenarios when one client authenticates on behalf of the other only.
+    /// </summary>
+    public class ClientAuthenticationForEdgeHubOnBehalfOf : ClientAuthenticationWithSharedAccessKeyRefresh
+    {
+        /// <summary>
+        /// Creates an instance of this class.
+        /// </summary>
+        /// <param name="sharedAccessKey">Shared access key value for the higher-layer parent device.</param>
```
Comment (patilsnr) [inline iothub/device/src/Authentication/ClientAuthenticationForEdgeHubOnBehalfOf.cs:18]: I'd love a double check on the correctness of these parameter descriptions, as I'm still getting familiar with this form of authentication

File: iothub/device/tests/Authentication/EdgeDeviceOnBehalfOfTests.cs [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    public class EdgeDeviceOnBehalfOfTests
+    {
+        private static string _testKey => Convert.ToBase64String(Encoding.UTF8.GetBytes(Guid.Empty.ToString("N")));
+
+        [TestMethod]
+        public async Task ConnectDeviceOnBehalfOf()
+        {
+            string leafDeviceId = "test-leaf-device";
+
+            DeviceConnectionSettings edgeDevice = new DeviceConnectionSettings("e4k-hub.azure-devices.net", "test-edge-device")
```
Comment (patilsnr) [inline iothub/device/tests/Authentication/EdgeDeviceOnBehalfOfTests.cs:20]: `DeviceConnectionSettings` and `EdgeHubConnectionSettingsBuilder` are classes that don't exist yet -- I'm looking for guidance on whether to add them from the e4k connector codebase or to find analogous places here in V2 to replace their functionality for these tests

File: iothub/device/src/Authentication/ClientAuthenticationForEdgeHubOnBehalfOf.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.Devices.Client.Authentication
+{
+    /// <summary>
+    /// Authentication method that generates shared access signature (SAS) token with refresh, based on a provided shared access key (SAK).
+    /// Build for gateway scenarios when one client authenticates on behalf of the other only.
```
Comment (rido-min) [inline iothub/device/src/Authentication/ClientAuthenticationForEdgeHubOnBehalfOf.cs:11]: Can we mention in the comments that this only works for $edgeHub ?

`gateway scenarios` can be confusing, I think we mean "when using the $edgeHub in IoTEdge to authenticate on behalf of leaf devices or modules"

File: iothub/device/src/Authentication/ClientAuthenticationForEdgeHubOnBehalfOf.cs (hunk 1 above)
Comment (rido-min) [inline iothub/device/src/Authentication/ClientAuthenticationForEdgeHubOnBehalfOf.cs:18]: The SharedAccessKey is the key for `$edgehub` module

File: iothub/device/src/Authentication/Security/SharedAccessSignatureBuilder.cs [hunk 4]
```diff
@@ -145,5 +145,30 @@ internal static string Sign(string requestString, string key)
             using var algorithm = new HMACSHA256(Convert.FromBase64String(key));
             return Convert.ToBase64String(algorithm.ComputeHash(Encoding.UTF8.GetBytes(requestString)));
         }
+
+        internal static string GetSignature(string encodedURI, string key, long expiry)
+        {
+            return Sign($"{encodedURI}\n{expiry}", key);
+        }
+
+        internal static string GetToken(string encodedURI, string key, long expiry = 0, int defaultTimeToLive = 60)
```
Comment (rido-min) [inline iothub/device/src/Authentication/Security/SharedAccessSignatureBuilder.cs:154]: Seems there is some duplication with existing methods in this class, so now there are two mechanisms to create the tokens. 

Any reason to do not use BuildSignature()?

File: iothub/device/tests/Authentication/EdgeDeviceOnBehalfOfTests.cs (hunk 2 above)
Comment (rido-min) [reply iothub/device/tests/Authentication/EdgeDeviceOnBehalfOfTests.cs:20]: I would not include the DeviceConnectionSettings in V2. The class is used just to hold the connection settings, you should be able to use IotHubConnectionString or just the individual values required to establish the connection

File: iothub/device/tests/Authentication/EdgeDeviceOnBehalfOfTests.cs [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+namespace Microsoft.Azure.Devices.Client.Tests.OnBehalfOf
+{
+    public class EdgeDeviceOnBehalfOfTests
+    {
+        private static string _testKey => Convert.ToBase64String(Encoding.UTF8.GetBytes(Guid.Empty.ToString("N")));
+
+        [TestMethod]
+        public async Task ConnectDeviceOnBehalfOf()
+        {
+            string leafDeviceId = "test-leaf-device";
```
Comment (rido-min) [inline iothub/device/tests/Authentication/EdgeDeviceOnBehalfOfTests.cs:18]: wrt test device ids: 
are you using the same approach for e2e tests? (assume the device identity has been provisioned and we have a known parent child relationship)

File: iothub/device/tests/Authentication/EdgeDeviceOnBehalfOfTests.cs [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+            string edgeModuleId = "test-leaf-device-module";
+
+            DeviceConnectionSettings edgeDevice = new DeviceConnectionSettings("e4k-hub.azure-devices.net", "test-edge-device")
+            {
+                SharedAccessKey = _testKey
+            };
+
+            var edgeHubCs = await EdgeHubConnectionSettingsBuilder.Build(edgeDevice);
```
Comment (rido-min) [inline iothub/device/tests/Authentication/EdgeDeviceOnBehalfOfTests.cs:68]: Where is EdgeHubConnectionSettingsBuilder? bringing this feature to V2 is very compelling (to enable provision any module, and initialize edgeHub), but I would leave it for a different PR.

@abhipsaMisra  ?

Comment (rido-min) [reply iothub/device/tests/Authentication/EdgeDeviceOnBehalfOfTests.cs:68]: btw.. the tests can be initialized with a _known_  $edgehub key

File: iothub/device/tests/Authentication/EdgeDeviceOnBehalfOfTests.cs [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+                edgeHubCs.SharedAccessKey!,
+                edgeHubCs.DeviceId!,
+                leafDeviceId,
+                null,
+                TimeSpan.FromMinutes(10),
+                5);
+
+            IotHubDeviceClient leafClient = new(edgeHubCs.HostName, leafAuth,
+                new IotHubClientOptions(
+                    new IotHubClientAmqpSettings
```
Comment (rido-min) [inline iothub/device/tests/Authentication/EdgeDeviceOnBehalfOfTests.cs:37]: should we test with MQTT as well?

we dont need to configure multiplexing in any test

**Review by rido-min — CHANGES_REQUESTED** [review]
- Clarify some comments
- Avoid bringing more classes from SdkLite
- We need a test strategy to provision required parent/child devices

File: iothub/device/tests/Authentication/EdgeDeviceOnBehalfOfTests.cs [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+            await leafClient.DisposeAsync();
+        }
+
+        [TestMethod]
+        public async Task ConnectLeafModuleOnBehalfOf()
+        {
+            string leafDeviceId = "test-leaf-device";
+            string edgeModuleId = "test-leaf-device-module";
+
+            var edgeHubCs = new IotHubConnectionString("e4k-hub.azure-devices.net", null, "test-edge-device", "$edgeHub", null, _testKey, null);
```
Comment (andyk-ms) [inline iothub/device/tests/Authentication/EdgeDeviceOnBehalfOfTests.cs:60]: is there a reason why "test-leaf-device" is in a variable and "test-edge-device" is not?

File: iothub/device/tests/Authentication/EdgeDeviceOnBehalfOfTests.cs [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+            twin.Should().NotBeNull();
+            twin.Reported.Should().NotBeNull();
+            twin.Reported["tick"].Should().NotBeNull();
+            tick.Should().Be((long)twin.Reported["tick"]);
+            await leafClient.CloseAsync();
+            await leafClient.DisposeAsync();
+        }
+
+        [TestMethod]
+        public async Task ConnectLeafModuleOnBehalfOf()
```
Comment (andyk-ms) [inline iothub/device/tests/Authentication/EdgeDeviceOnBehalfOfTests.cs:56]: Would it be better to consolidate two tests to call a single test function with different inputs for easier maintenance given both tests are similar?
